### PR TITLE
http2: request.socket|connection

### PR DESCRIPTION
**Pull request by sebdeckers**
As discussed in: https://github.com/nodejs/http2/pull/125#discussion_r116943863

Adds `request.socket` and `request.connection` for compatibility with documented HTTP APIs:

- > message.socket
  > Added in: v0.3.0
  > <net.Socket>
  > The net.Socket object associated with the connection.
  > https://nodejs.org/api/http.html#http_message_socket
- > Event: 'connection'
  > [...]
  > The socket can also be accessed at request.connection.
  > https://nodejs.org/api/http.html#http_event_connection

Use case: I found that popular packages like [cookies](https://npmjs.org/package/cookies) (NPM: >500k/mo) [make use of this to detect the TLS context](https://github.com/pillarjs/cookies/blob/master/index.js#L87).

##### Checklist

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-guidelines)

##### Affected core subsystem(s)

http2

### CODE REVIEW

**Review by mcollina — CHANGES_REQUESTED** [review]
Can you add a test that verifies how this is used by `cookies`?

**Comment by sebdeckers** [comment]
@mcollina Do you mean just replicate that logic in our test, or to actually pull in the `cookies` package from NPM and use it in a test? (If the latter, what is the recommended way to add 3rd party packages to the node repo?)

**Comment by mcollina** [comment]
We need a line that just does this: https://github.com/pillarjs/cookies/blob/master/index.js#L87

**Comment by sebdeckers** [comment]
Okay, I'm just doing a look around for other examples.

#### finalhandler (NPM: 14M/mo)

> ```js
> req.socket.destroy()
> ```
> https://github.com/pillarjs/finalhandler/blob/master/index.js#L130

#### express (NPM: ∞/mo)

> ```js
> req.connection.encrypted = true;
> ```
> https://github.com/expressjs/express/blob/master/test/req.protocol.js#L41
> ```js
> var proto = this.connection.encrypted
>   ? 'https'
>   : 'http';
> var trust = this.app.get('trust proxy fn');
>
> if (!trust(this.connection.remoteAddress, 0)) {
>   return proto;
> }
> ```
> https://github.com/expressjs/express/blob/master/lib/request.js#L307-L314

**Comment by mcollina** [comment]
https://github.com/pillarjs/finalhandler/blob/master/index.js#L127-L132 cannot be supported as HTTP1 specific. In that specific case for HTTP1, we would have to destroy the stream, not the connection.

`finalhandler` also uses `res._header`  which was not documented in core. The one we should support is https://nodejs.org/api/http.html#http_response_headerssent. I will send a PR to `finalhandler`.

**Comment by sebdeckers** [comment]
#### hapi

```js
    // Setup timeout

    if (this.raw.req.socket &&
        this.route.settings.timeout.socket !== undefined) {

        this.raw.req.socket.setTimeout(this.route.settings.timeout.socket || 0);    // Value can be false or positive
    }
```

https://github.com/hapijs/hapi/blob/master/lib/request.js#L344-L350

**Comment by mcollina** [comment]
I'm convinced. We need this :).

In the test, can you check that whatever `socket`  returns is correct? Just asserting its truthness is not enough.

**Comment by sebdeckers** [comment]
Added a simple `typeof` check which should cover all the public interfaces in use today.

Should I add a test case for HTTPS too? There's not really a difference in core/compat (both are just `Http2Session`). And tls.TLSSocket is a subclass of net.Socket anyway.

**Comment by mcollina** [comment]
that's fine as it is

**Review by mcollina — APPROVED** [review]
LGTM

**Review by cjihrig — APPROVED** [review]

File: lib/internal/http2/compat.js [hunk 1]
```diff
@@ -181,6 +181,14 @@ class Http2ServerRequest extends Readable {
     return '2.0';
   }
 
+  get socket() {
+    return this.stream.session.socket;
+  }
+
+  get connection() {
+    return this.socket;
```
Comment (jasnell) [inline lib/internal/http2/compat.js:189]: One point on this: we *really* need to discourage direct access to the socket. With an http/2 connection, bad things can happen if the user does not know what they are doing. The underlying native Http2Session class takes complete ownership of reading and writing to the socket and any direct interactions with the socket by user code could throw the entire connection out of sync. I'd prefer not to expose these at all but if we need to, let's add a strong warning in the docs.

Comment (sebdeckers) [reply lib/internal/http2/compat.js:189]: Okay, will rebase and update the docs that were just landed.

**Comment by jasnell** [comment]
Regarding the hapi setTimeout example... I'm curious why it is necessary actually. In the http/2 case, the `Http2Session` instance itself has a `setTimeout()` and there is a 1-to-1 relationship between the Socket and the Session. Setting the timeout on the `Http2Session` should have the same basic effect, then. In fact, it may be better to accept have `setTimeout()` on `Http2Session` defer to the `setTimeout()` on the socket rather than it being a separate implementation.

**Comment by sebdeckers** [comment]
@jasnell I suspect it stems from wanting to set different timeouts for specific Hapi routes. See: https://github.com/hapijs/hapi/issues/2195 It would only be possible to apply the timeout once the request has been parsed, not at the initial socket connection listener. Just a guess.

**Comment by jasnell** [comment]
Yeah, I get that, but given the multiplexing model, and the fact that individual `Http2Stream` objects have their *own* timeouts that are independent of the `Http2Session` and `Socket`, does that existing model still make sense here?

File: doc/api/http2.md [hunk 2]
```diff
@@ -227,6 +227,8 @@ process has completed.
 A reference to the [`net.Socket`][] or [`tls.TLSSocket`][] to which this
 `Http2Session` instance is bound.
 
+Directly reading from and writing to the socket is strongly discouraged.
```
Comment (jasnell) [inline doc/api/http2.md]: Likely need to take the time to explain why, but this could be a whole section on its own.

**Comment by mcollina** [comment]
Landed as 3904d7c6c142fe492699f7d4d07f4328b1691720

**Pull request closed without merging**